[JENKINS-42950] Support more credential masking scenarios

jvz opened the pull request:
[JENKINS-42950](https://issues.jenkins-ci.org/browse/JENKINS-42950)

This adds a new extension point, `SecretPatternFactory`, that can be used to implement different quoting and unquoting strategies used by shells that can mangle passwords and other secrets in their output. This is used to improve credential masking in more scenarios where the secret was not intended to be seen.

Co-Authored-By: @Wadeck 

@reviewbybees @jeffret-b @jglick @jtnord @daniel-beck

daniel-beck commented on src/main/resources/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep/help.html:
```diff
@@ -73,13 +73,8 @@
-<p>
-    This particular issue can be worked around by turning off echo with <code>set +x</code>,
-    or simply avoiding possible shell metacharacters in secrets.
+    Mangled secrets can only be detected on a best effort basis using the <code>PatternMaskerProvider</code>
```
Not something I would include in user documentation: Nobody cares how it's done.

daniel-beck commented on src/main/resources/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep/help.html:
```diff
@@ -73,13 +73,8 @@
-    This particular issue can be worked around by turning off echo with <code>set +x</code>,
-    or simply avoiding possible shell metacharacters in secrets.
+    Mangled secrets can only be detected on a best effort basis using the <code>PatternMaskerProvider</code>
+    extension point. Default providers include one for handling common ways that data may be encoded by Bash and Batch.
```
The previous docs actually did a better job explaining the problem. The new version just mentions "mangled secrets" without giving users an idea what those are.

jvz replied:
I've updated the docs to retain some of the more useful info I had removed.

daniel-beck commented on src/main/resources/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep/help.html:
```diff
@@ -73,13 +73,8 @@
-    This particular issue can be worked around by turning off echo with <code>set +x</code>,
-    or simply avoiding possible shell metacharacters in secrets.
+    Mangled secrets can only be detected on a best effort basis using the <code>PatternMaskerProvider</code>
+    extension point. Default providers include one for handling common ways that data may be encoded by Bash and Batch.
```
```suggestion
    extension point. By default, Jenkins will attempt to mask mangled secrets as they would appear in output of Bash and Windows batch.
```

Wadeck commented on src/test/java/org/jenkinsci/plugins/credentialsbinding/impl/CredentialsMaskingBashTest.java:
```diff
@@ -0,0 +1,118 @@
+
+@RunWith(Theories.class)
+@For(BashPatternMaskerProvider.class)
+public class CredentialsMaskingBashTest {
```
All the `@For` should be removed by simply renaming the test with standard names, like here `BashPAtternMaskerProviderTest`

Wadeck commented on src/main/java/org/jenkinsci/plugins/credentialsbinding/masking/BatchPatternMaskerProvider.java:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    @Override
+    @Nonnull String getUnquotedForm(@Nonnull String input) {
```
⚠️  I replaced all the other method's bodies by just `return input` and all the tests were passing. Please either remove the `surroundWithQuotes` / `getQuotedForm` content or add test that demonstrate their usage. 
Otherwise we are creating brand new dead code.

jvz replied:
Added a test, let's see if that's demonstrative or not.

Wadeck replied:
@jvz I redo the same with the new test, same results. None of the method brings something in that case.

jeffret-b replied:
I'm not sure I'm following this comment. Is it specific to the Batch implementation? Making the suggested changes in the Shell implementation results in failed tests.

jvz replied:
It seems like my attempts at unifying batch and bash escaping APIs may be overkill if batch never has this problem. Which makes me wonder if anyone has ever actually reported a bug around this for batch in the first place.

jeffret-b replied:
Good point.

Wadeck replied:
Reducing the interface to just one method, asking for a set/list of alternative format could be sufficient I imagine.

The ^ case could be easily be a real-life scenario, just I am not aware of any reported issue especially on that area.

Wadeck commented on src/main/java/org/jenkinsci/plugins/credentialsbinding/masking/PatternMaskerProvider.java:
```diff
@@ -0,0 +1,77 @@
+
+    @Restricted(NoExternalUse.class)
+    static @Nonnull Collection<String> getAllAlternateForms(@Nonnull String input) {
+        Collection<String> alternateForms = new TreeSet<>(BY_REVERSE_LENGTH);
```
At this point, a basic `HashSet` is sufficient. For the reader it's harder to understand why you put a tree here and in the next method, while only the latter one is required.

jvz replied:
Probably an over optimization since the set sorting doesn't need to happen until the regex is created.

Wadeck commented on src/main/java/org/jenkinsci/plugins/credentialsbinding/masking/PatternMaskerProvider.java:
```diff
@@ -0,0 +1,77 @@
+
+    @Restricted(NoExternalUse.class)
+    static @Nonnull Pattern getMaskingPattern(@Nonnull Collection<String> inputs) {
+        Collection<String> patterns = new TreeSet<>(BY_REVERSE_LENGTH);
```
You should make explicit the requirement to have this ordered collection (variable name, comment?). 
Two points to mention:

1. If you have two secrets with one being a substring of the other, there is no guarantee the longest will be masked completely (should be part of the documentation around limitation)
2. The reverse length order is used in case you have for a given secret multiple patterns of different length, like `abc` and `'abc'`, you do not want to have the shorter version used before the longer and thus, giving information about the password content in the log.

The second point is particularly interesting as it's not demonstrated in the test, but it should be from my PoV.

jvz replied:
My main use case for this ordering is so that `abc` and `'abc'` are not masked as `****` and `'****'` respectively. I try to include the surrounding quotes, though now I wonder if that's leaking information or not.

jvz replied:
Oh I see what you mean. I'll updated the javadoc here.

Wadeck commented on src/test/java/org/jenkinsci/plugins/credentialsbinding/impl/CredentialsMaskingBashTest.java:
```diff
@@ -0,0 +1,118 @@
+
+    @DataPoints
+    public static List<String> generatePasswords() {
+        ThreadLocalRandom random = ThreadLocalRandom.current();
```
Due to the non-deterministic RNG you are using, the tests could pass, then fail, then pass. I would prefer to have a fix seed to avoid such behavior (even at the cost of increasing the number of rounds)

Wadeck commented on src/test/java/org/jenkinsci/plugins/credentialsbinding/impl/CredentialsMaskingPowerShellTest.java:
```diff
@@ -0,0 +1,139 @@
+
+    @Before
+    public void assumeWindowsForBatch() throws Exception {
+        // TODO: pwsh is also a valid executable name
```
Perhaps add the reference to https://github.com/jenkinsci/durable-task-plugin/pull/88.

The people there seem to agree on having 2 different commands.

Wadeck commented on src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/BashPatternMaskerProviderTest.java:
```diff
@@ -0,0 +1,84 @@
+
+        WorkflowRun run = j.assertBuildStatusSuccess(project.scheduleBuild2(0));
+
+        j.assertLogContains(": ****", run);
```
How do you differentiate between the two?
In the case one is outputting `: foo\\bar\\` (FTR password = `foo\bar\` due to Java backslashes), the test will pass but in reality it should have failed.

jvz replied:
I'm updating this test to check that neither "foo" nor "bar" are present in the output.

jvz commented:
Alright, I think I've addressed all your feedback. Not sure why CI is failing as that's giving a different output than my own testing locally for bash.

jglick commented:
> Not sure why CI is failing as that's giving a different output than my own testing locally for bash.

Perhaps because the CI build is running on Ubuntu. Your test seems to assume without checking that `sh` is actually Bash, which it is on some less-compliant Linux distributions or platforms.

jvz commented:
> > Not sure why CI is failing as that's giving a different output than my own testing locally for bash.
> 
> Perhaps because the CI build is running on Ubuntu. Your test seems to assume without checking that `sh` is actually Bash, which it is on some less-compliant Linux distributions or platforms.

Good catch. I've updated the tests to change the default shell to bash.

jvz commented:
Now I'm confused as to what's failing this time.

Wadeck commented on src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/BashPatternMaskerProvider2Test.java:
```diff
@@ -52,6 +54,7 @@
     @Before
     public void setUp() {
         assumeThat("bash", is(executable()));
+        j.jenkins.getDescriptorByType(Shell.DescriptorImpl.class).setShell(Executables.getPathToExecutable("bash"));
```
Due to that, you remove the test for "pure Shell" right? But we want to support it along Bash, Batch and PowerShell.

jvz commented:
Non-bash implementations of Bourne shell do not perform any quoting of
echoed commands. It works more like Windows Batch where it just echoes out
what you typed. I don’t think it’s documented here, but I did find that
during discovery.

On Mon, May 6, 2019 at 06:50, Wadeck Follonier <notifications@github.com>
wrote:

> *@Wadeck* commented on this pull request.
> ------------------------------
>
> In
> src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/BashPatternMaskerProvider2Test.java
> <https://github.com/jenkinsci/credentials-binding-plugin/pull/59#discussion_r281152781>
> :
>
> > @@ -52,6 +54,7 @@
>      @Before
>      public void setUp() {
>          assumeThat("bash", is(executable()));
> +        j.jenkins.getDescriptorByType(Shell.DescriptorImpl.class).setShell(Executables.getPathToExecutable("bash"));
>
> Due to that, you remove the test for "pure Shell" right? But we want to
> support it along Bash, Batch and PowerShell.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jenkinsci/credentials-binding-plugin/pull/59#pullrequestreview-233938196>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGBF2ZOKHXXZMIRZYRALD3PUALP3ANCNFSM4HKFF2XA>
> .
>
-- 
Matt Sicker
Senior Software Engineer, CloudBees

Wadeck commented:
@jvz Re Shell => so a non-regression test can be written for such case easily

jvz commented:
@Wadeck: added tests for sh

jeffret-b commented on src/main/java/org/jenkinsci/plugins/credentialsbinding/masking/BashPatternMaskerProvider.java:
```diff
@@ -0,0 +1,64 @@
+
+    @Override
+    @Nonnull String surroundWithQuotes(@Nonnull String input) {
+        return "'" + input + "'";
```
The implementation of surroundWithQuotes() and getUnquotedForm() seems to be duplicated between the Bash and Batch classes. Does it make sense to eliminate that duplication somehow?

jvz replied:
Bash uses single quotes, batch uses double quotes. Not much to deduplicate.

jeffret-b replied:
Ah, I kept trying to detect the difference, but I missed that.

jeffret-b commented on src/test/java/org/jenkinsci/plugins/credentialsbinding/impl/CredentialsMaskingBatchTest.java:
```diff
@@ -0,0 +1,300 @@
+    }
+
+    @Test
+    @Ignore("Cannot support the dangerous characters in direct expressions")
```
What's the value of having these ignored tests?

jvz replied:
@Wadeck had initially authored them, probably to show all the possibilities?

Wadeck replied:
Yeah and to show the "impossibility" to support the dangerous stuff.

jeffret-b replied:
I don't think the ignored tests do a good job of showing that. It's basically introducing dead code that we never expect to fix. I'd prefer a different approach. One possibility would be to remove the `@Ignore`s, implement the test to expect the known behavior, and then leave some comments or naming to indicate that while this isn't the desired behavior it is expected. At least this way the tests do something, affirming the behavior continues.

jvz replied:
I tried updating them to check for a failed build. Let's see how that works out.

Wadeck replied:
@jvz  just in case, I sent you the patch to apply that modification (https://cloudbees.slack.com/archives/DBZE45TCM/p1557400247010700)

jvz commented:
TODO:

- [ ] Simplify batch escaped forms generation
- [ ] Remove or modify ignored tests to be more useful

jvz commented:
@Wadeck FYI, I rebased this PR using the patch you sent me.

daniel-beck commented on src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/BashPatternMaskerProviderTest.java:
```diff
@@ -0,0 +1,130 @@
+            int length = random.nextInt(24) + 8;
+            StringBuilder sb = new StringBuilder(length);
+            for (int j = 0; j < length; j++) {
+                char next = (char) (random.nextInt('~' - ' ' + 1) + ' '); // space = 0x20, tilde = 0x7E
```
This appears to exclude space characters. Deliberate?

jvz replied:
It's supposed to be the range from space to tilde inclusive on both sides.

jvz replied:
Oh I see what you're looking at. No, this range is created to include space to tilde inclusive so as to avoid writing the character literal for 0x7f.

daniel-beck replied:
I misread the code as the `+1` being outside the args of `#nextInt`.

And since the bound value (here `0x5f`) is exclusive, we never get there. All good.

jvz replied:
0x7f. And I updated the code to be a bit more clear by just using the upper bound directly and including the comment on printable ASCII ranges.

daniel-beck replied:
`0x5f` is the arg passed to `nextInt`, then you add `0x20` to the result to get the char.

jvz replied:
Oh right, got it.

daniel-beck commented on src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/BashPatternMaskerProviderTest.java:
```diff
@@ -0,0 +1,130 @@
+            int length = random.nextInt(24) + 8;
+            StringBuilder sb = new StringBuilder(length);
+            for (int j = 0; j < length; j++) {
+                char next = (char) (random.nextInt('~' - ' ' + 1) + ' '); // space = 0x20, tilde = 0x7E
```
```suggestion
                char next = (char) (random.nextInt('~' - ' ' + 1) + ' '); // space/0x20 is the first printable ASCII char, tilde/0x7E the last
```

daniel-beck commented on src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/BashPatternMaskerProviderTest.java:
```diff
@@ -0,0 +1,130 @@
+public class BashPatternMaskerProviderTest {
+
+    public static final @DataPoint String SAMPLE_PASSWORD = "}#T14'GAz&H!{$U_";
+    public static final @DataPoint String ANOTHER_SAMPLE_PASSWORD = "a'b\"c\\d(e)#";
```
These passwords, while containing shell metacharacters, don't appear to be values chosen specifically to break escaping. Something like `'"'(foo)'"'` might be more convincing IMO.

jvz replied:
Added your example.

daniel-beck commented on src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/BourneShellPatternMaskerProviderTest.java:
```diff
@@ -0,0 +1,127 @@
+import static org.junit.Assume.assumeThat;
+
+@RunWith(Theories.class)
+public class BourneShellPatternMaskerProviderTest {
```
Lots of code duplication between this and the `Bash…Test`

jvz replied:
I blame JUnit 4. This would be easier to structure in 5.x.

jvz replied:
Cleaned up code duplication between all the pattern masking tests.

jvz commented on src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/BatchPatternMaskerProviderTest.java:
```diff
@@ -0,0 +1,377 @@
+    private void setupProject(String pipeline) throws Exception {
+        String projectName = UUID.randomUUID().toString();
+        project = j.jenkins.createProject(WorkflowJob.class, projectName);
+        credentialId = UUID.randomUUID().toString();
```
This seems to conflict with the `credentialId` set by `registerCredentials()`

jvz commented:
Ugh, something broke in batch escaping. Need to fix that.

jvz commented:
Looks like I forgot to call `Pattern.quote()` in the refactored batch masker. Pushed an update.

jvz commented:
Can I get one of the maintainers to merge? @abayer @jglick @jtnord @batmat @Vlatombe @stephenc

jglick commented:
That would be me. I have not yet had time to review.

jglick commented:
See also #55.

jglick commented on src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/BashPatternMaskerProvider2Test.java:
```diff
@@ -0,0 +1,76 @@
+        String credentialsId = CredentialsTestUtil.registerStringCredentials(j.jenkins, password);
+        project.setDefinition(new CpsFlowDefinition(
+                "node {\n" +
+                        "  withCredentials([string(credentialsId: '" + credentialsId + "', variable: 'CREDENTIALS')]) {\n" +
```
```suggestion
                "  withCredentials([string(credentialsId: '" + credentialsId + "', variable: 'CREDENTIALS')]) {\n" +
```

to make scripts more readable

jglick commented on src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/BashPatternMaskerProvider2Test.java:
```diff
@@ -0,0 +1,76 @@
+                "node {\n" +
+                        "  withCredentials([string(credentialsId: '" + credentialsId + "', variable: 'CREDENTIALS')]) {\n" +
+                        "    sh ': $CREDENTIALS'\n" + // forgot quotes
+                        "    sh \": $CREDENTIALS\"\n" + // using groovy variable and forgot quotes
```
```suggestion
                        "    sh(/: $CREDENTIALS/)\n" + // using Groovy variable and forgot quotes
```

for readability (avoiding `\"` in Java sources)

jglick commented on src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/BashPatternMaskerProvider2Test.java:
```diff
@@ -0,0 +1,76 @@
+        j.assertLogContains(": ****", run);
+        j.assertLogNotContains(password, run);
+        j.assertLogNotContains("foo", run);
+        j.assertLogNotContains("bar", run);
```
Maybe use more distinctive password components to avoid false failures due to unrelated text, like some hypothetical

```
Running Pipeline now, barring any further delays
```

jglick commented on src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/BashPatternMaskerProvider2Test.java:
```diff
@@ -0,0 +1,76 @@
+import static org.jenkinsci.plugins.credentialsbinding.test.Executables.executable;
+import static org.junit.Assume.assumeThat;
+
+public class BashPatternMaskerProvider2Test {
```
```suggestion
@For(BashPatternMaskerProvider.class)
public class BashPatternMaskerProvider2Test {
```

jglick commented on src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/BashPatternMaskerProviderTest.java:
```diff
@@ -0,0 +1,121 @@
+
+    @DataPoints
+    public static List<String> generatePasswords() {
+        Random random = new Random(100);
```
(thank you for using a deterministic seed)

jglick commented on src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/BashPatternMaskerProviderTest.java:
```diff
@@ -0,0 +1,121 @@
+            int length = random.nextInt(24) + 8;
+            StringBuilder sb = new StringBuilder(length);
+            for (int j = 0; j < length; j++) {
+                char next = (char) (' ' + random.nextInt('\u007f' - ' ')); // 0x7f is DEL, 0x7e is ~, and space is the first printable ASCII character
```
So this is in the range `[' ', '~']` IIUC?

jvz replied:
Yes.

jglick commented on src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/BashPatternMaskerProviderTest.java:
```diff
@@ -0,0 +1,121 @@
+        return passwords;
+    }
+
+    @Rule public JenkinsRule j = new JenkinsRule();
```
Do you want to make this a `@ClassRule`, or do we already only start Jenkins once for the suite? The [CI results](https://ci.jenkins.io/job/Plugins/job/credentials-binding-plugin/job/PR-59/17/testReport/org.jenkinsci.plugins.credentialsbinding.masking/BashPatternMaskerProviderTest/linux_8___Archive__linux_8____credentialsAreMaskedInLogs/) show only 20s so it is unclear.

jglick replied:
Checked and it does start Jenkins each time; `@ClassRule` fixes that. (Generally it is tricky to use, but you are already using `JenkinsRule.createProject(Class)` and `UUID.randomUUID()` to ensure that each test case uses distinct identifiers.)

jglick commented on src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/BashPatternMaskerProviderTest.java:
```diff
@@ -0,0 +1,121 @@
+        assumeThat("bash", is(executable()));
+        // due to https://github.com/jenkinsci/durable-task-plugin/blob/e75123eda986f20a390d92cc892c3d206e60aefb/src/main/java/org/jenkinsci/plugins/durabletask/BourneShellScript.java#L149
+        // on Windows
+        assumeThat("nohup", is(executable()));
```
Why not just `assumeFalse(Functions.isWindows())`?

jvz replied:
I think this is a workaround to make bash work on Windows in pipelines.

jglick commented on src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/BashPatternMaskerProviderTest.java:
```diff
@@ -0,0 +1,121 @@
+                "node {\n" +
+                        "  withCredentials([string(credentialsId: '" + credentialsId + "', variable: 'CREDENTIALS')]) {\n" +
+                        "    sh ': \"$CREDENTIALS\"'\n" + // : will expand its parameters and do nothing with them
+                        "    sh ': \"< $CREDENTIALS >\"'\n" +
```
What is the significance of the `<`…`>` here? (a comment would be in order)

jvz replied:
Oh, nothing more significant than using "before1" and "after1", just less typing.

jvz replied:
Adding docs. Other resolved things are likewise being addressed.

jglick commented on src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/BatchPatternMaskerProviderTest.java:
```diff
@@ -0,0 +1,367 @@
+
+    @Test
+    public void simple_singleQuote() throws Exception {
+        registerCredentials(SIMPLE);
```
Why is the Bash version using the theory system, and this is done manually? Smells like there could be some consolidation of test code.

jvz replied:
Different test authors originally. I can likely figure out how to break this into a theorized test.

jglick commented on src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/BatchPatternMaskerProviderTest.java:
```diff
@@ -0,0 +1,367 @@
+        assertDelayedNoPlainTextButStars(runDelayedAllQuotes());
+    }
+
+    // we do NOT support dangerous characters in direct expansion mode
```
Sorry, what is “direct expansion mode” and what is not supported? `help.html` mentions nothing of the sort.

daniel-beck replied:
I think this refers to what `setupNoQuoteProject` does.

jglick replied:
I am still unclear on what “direct expansion mode” means, or what this means:

> these commands would be completely broken if echo didn't work the way it does when CREDENTIALS contains special characters

What is the way that echo does work, or how else would it work, and what commands would be “broken” and how?

jvz replied:
Direct expansion is the default mode; delayed expansion is the newer cmd feature that defers expansion of environment variables until the command is executed instead of while it's parsed.

jglick commented on src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/BatchPatternMaskerProviderTest.java:
```diff
@@ -0,0 +1,367 @@
+        setupProject("node {\n" +
+                "  withCredentials([string(credentialsId: '" + credentialId + "', variable: 'CREDENTIALS')]) {\n" +
+                "    bat \"\"\"\n" +
+                "      echo before1 $CREDENTIALS after1\n" + // DO NOT DO THIS IN PRODUCTION; IT IS USING THE WRONG VARIABLE SYNTAX
```
```suggestion
                "      echo before1 $CREDENTIALS after1\n" + // DO NOT DO THIS IN PRODUCTION; IT IS USING GROOVY INTERPOLATION
```

or WDYM?

jglick commented on src/test/java/org/jenkinsci/plugins/credentialsbinding/test/Executables.java:
```diff
@@ -0,0 +1,65 @@
+import java.util.List;
+
+public class Executables {
+    private static final String LOCATOR = Functions.isWindows() ? "where.exe" : "which";
```
or just use, e.g.,

```groovy
sh '''#!/usr/bin/env bash -ex
# as before…
'''
```

jvz replied:
This is also used for determining if the machine has bash available rather than just assuming the environment.

jvz commented on src/main/java/org/jenkinsci/plugins/credentialsbinding/masking/PatternMaskerProvider.java:
```diff
@@ -0,0 +1,95 @@
+     * For example, {@code bash -x} will only quote arguments echoed when necessary. To avoid leaking the presence or
+     * absence of quoting, the longer form is masked.
+     */
+    @Restricted(NoExternalUse.class)
```
What type of secrets?

jglick replied:
Forget about it, not important here.

jvz commented on src/main/java/org/jenkinsci/plugins/credentialsbinding/masking/PatternMaskerProvider.java:
```diff
@@ -0,0 +1,95 @@
+
+    /**
+     * Returns a collection of alternative forms as regular expressions the given input may show up as in logs. These
+     * patterns should be appropriately {@linkplain Pattern#quote(String) quoted} for matching literal patterns as
```
Alright, I refactored this a bit and renamed the interface to SecretPatternFactory.

jglick commented on src/main/java/org/jenkinsci/plugins/credentialsbinding/masking/SecretPatternFactory.java:
```diff
@@ -0,0 +1,61 @@
+     * Returns a collection of alternative forms the given input may show up as in logs.
+     * Note that these patterns must embed their flags in the pattern rather than as parameters to Pattern.
+     */
+    @Nonnull Collection<Pattern> getSecretPatterns(@Nonnull String input);
```
This still bakes the use of regexps into the SPI. Given that all the known implementations just produce a collection of literal strings to be masked, we could start with a simpler SPI that just lets you return a set of such strings. That would allow the use of an aggregated regexp to be a purely internal implementation detail.

If you restrict the extension point, then we always have the option of making a powerful SPI later that allows an implementation to mask a whole class of related strings. For that matter, even if you did _not_ restrict it, we could simply add another extension point and consult both. But I do not think there is any actual use case for the set of implementations to be extended from another plugin, is there? Assuming there is no plan for any implementations outside this plugin, it would be better restricted.

jvz commented:
I've addressed most of your review comments, @jglick. I'll take a look at improving the test differences later today.

jglick commented on src/main/java/org/jenkinsci/plugins/credentialsbinding/masking/SecretPatterns.java:
```diff
@@ -0,0 +1,69 @@
+
+    private static final Comparator<Pattern> EXTRACT_PATTERN = Comparator.comparing(Pattern::pattern);
+    private static final Comparator<Pattern> BY_LENGTH_DESCENDING = Comparator.comparing(Pattern::pattern,
+            Comparator.comparingInt(String::length).reversed().thenComparing(String::compareTo));
```
This is another reason to keep regexps out of the SPI: you are sorting by _pattern_ length, not length of masked strings per se. Assuming there is a good reason to sort to begin with (I cannot recall offhand the original motivating bug report), it could be dangerous to do it this way, since there is no guarantee that a textually long regexp matches longer strings than a textually shorter regexp.

jvz replied:
Yes, great point. I'll clean this up and give another whack at it.

jvz commented:
Updating the log message has tests? Color me surprised, ok. I'll also have a fix for that.

jglick commented on src/test/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStepTest.java:
```diff
@@ -146,7 +146,7 @@
                 WorkflowRun b = p.scheduleBuild2(0).waitForStart();
                 SemaphoreStep.waitForStart("basics/1", b);
-                story.j.assertLogContains(Functions.isWindows() ? "Masking only exact matches of %USERNAME% or %PASSWORD%" : "Masking only exact matches of $USERNAME or $PASSWORD", b);
+                story.j.assertLogContains(Functions.isWindows() ? "Masking supported pattern matches of %USERNAME% or %PASSWORD%" : "Masking supported pattern matches of $USERNAME or $PASSWORD", b);
```
Feel free to introduce a package-private utility method for this message to be shared between `src/main/` and `src/test/`.

jglick commented on src/main/java/org/jenkinsci/plugins/credentialsbinding/masking/BashSecretPatternFactory.java:
```diff
@@ -0,0 +1,72 @@
+        patterns.add(quotedForm);
+        patterns.add(surroundWithQuotes(quotedForm));
+        patterns.add(getUnquotedForm(input));
+        return patterns;
```
or a bit more simply

```java
return Arrays.asList(quotedForm, surroundWithQuotes(quotedForm), getUnquotedForm(input));
```

jglick commented on src/main/resources/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep/help.html:
```diff
@@ -73,13 +73,15 @@
-    Mangled secrets are not detected by Jenkins and will appear in plain text in the log file, in this case as:
+    Mangled secrets can only be detected on a best-effort basis. By default, Jenkins will attempt to mask mangled
+    secrets as they would appear in output of Bash and Windows batch. Without these strategies in place, mangled secrets
+    would appear in plain text in log files. In the example above, this would result in:
```
BTW for natural-language text files it is a good idea to use one physical line per sentence or major phrase. Easier to review & edit.
